File: generated_code/python/solution_1411_3.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def getDecimalValue(self, head: Optional[ListNode]) -> int:
        """
        Algorithm: Bitwise Shift / Horner's Method
        Time Complexity: O(N) where N is the number of nodes. We traverse the list exactly once.
        Space Complexity: O(1) as we only use a single integer variable to store the result.
        
        Since the most significant bit is at the head, each time we move to the next node,
        we can shift our current result to the left by 1 (equivalent to multiplying by 2)
        and then add the current node's value using a bitwise OR operation.
        """
        ans = 0
        current = head
        
        while current:
            # Shift the current value to the left and add the new bit
            ans = (ans << 1) | current.val
            current = current.next
            
        return ans

File: generated_code/cpp/solution_1411_3.cpp
#include

// Definition for singly-linked list if not already defined externally.
// Included here to ensure the code is self-contained and compilable.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    int getDecimalValue(ListNode* head) {
        // Time Complexity: O(N) where N is the number of nodes in the linked list.
        // Space Complexity: O(1) as we only use a single integer to accumulate the result.
        int result = 0;

        // Traverse the linked list from head to tail.
        // Since the head represents the most significant bit, each step shifts 
        // the current accumulated value to the left by 1 (equivalent to multiplying by 2)
        // and adds the current node's value using a bitwise OR operation.
        while (head != nullptr) {
            result = (result << 1) | head->val;
            head = head->next;
        }
        
        return result;
    }
};